diff --git a/CMakeLists.txt b/CMakeLists.txt
index 991fa21..05a5547 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -697,7 +697,7 @@ if (BUILD_WITH_TM2)
     )
 endif()
 
-set_source_files_properties(src/image_avx.cpp PROPERTIES COMPILE_FLAGS -mavx2)
+# set_source_files_properties(src/image_avx.cpp PROPERTIES COMPILE_FLAGS -mavx2)
 
 option(BUILD_SHARED_LIBS "Build shared library" ON)
 if(BUILD_SHARED_LIBS)
